Assets/Src/AuthController/CredentialProviders/Google/GoogleCredentialProvider.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System.Threading.Tasks;
+using Src.AuthController.AuthKeys;
+using Src.AuthController.REST;
+using Src.AuthController.REST.REST_Request_Proxies;
+using Src.AuthController.REST.REST_Response_DTOs;
+using UnityEngine;
+
+namespace Src.AuthController.CredentialProviders.Google
+{
+    public class GoogleCredentialProvider
+    {
+        private const int LoopbackPort = 3303;
+        private readonly string _redirectUri = $"http://localhost:{LoopbackPort}";
+        
+        private string _authCode;
+        
+        private GoogleIdTokenResponse _googleApiResponse;
+        private float _googleApiResponseIssueTime;
+        private const float AccessTokenValidityMargin = 30f;
+        
+        public virtual async Task<GoogleIdTokenResponse> GetCredentialAsync()
+        {
+            if (_googleApiResponse != null)
+            {
+                if (ValidateAccessToken()) 
+                {
+                    return _googleApiResponse;
+                }
+
+                var responseTcs = new TaskCompletionSource<GoogleRefreshTokenResponse>();
+                
+                RefreshAccessToken(responseTcs);
+                
+                await responseTcs.Task;
+                var response = responseTcs.Task.Result;
+
+                _googleApiResponse.access_token = response.access_token;
+                _googleApiResponse.expires_in = response.expires_in;
+                
+                return _googleApiResponse;
+            }
+            else
+            {
+                var responseTcs = new TaskCompletionSource<GoogleIdTokenResponse>();
+                
+                GetAuthData(responseTcs);
+                
+                await responseTcs.Task;
+                _googleApiResponse = responseTcs.Task.Result;
+                return _googleApiResponse;
+            }
+        }
+
+        private bool ValidateAccessToken()
+        {
+            int.TryParse(_googleApiResponse.expires_in, out var validityPeriod);
+
+            return (Time.time - _googleApiResponseIssueTime) <
+                   (validityPeriod - AccessTokenValidityMargin);
+        }
+        
+        private void GetAuthData(TaskCompletionSource<GoogleIdTokenResponse> tcs)
+        {
+            Application.OpenURL($"https://accounts.google.com/o/oauth2/v2/auth?client_id=" +
```

Comment:
Move logic that is hard to test into separate classes and use them via interfaces. Consider using Humble object pattern.